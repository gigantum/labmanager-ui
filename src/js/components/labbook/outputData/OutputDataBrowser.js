// vendor
import React, { Component } from 'react'
import {createPaginationContainer, graphql} from 'react-relay'
//Config
import FileBrowserWrapper from 'Components/labbook/fileBrowser/FileBrowserWrapper'
//store
import store from 'JS/redux/store'

class OutputDataBrowser extends Component {
  constructor(props){
  	super(props);
    const {owner, labbookName} = store.getState().routes
    this.state = {
      'show': false,
      'message': '',
      'files': [],
      'moreLoading': false,
      owner,
      labbookName
    }

  }

  /*
    loads more if branches are switched
  */
  componentDidUpdate(){
    if(!this.state.moreLoading && this.props.output.allFiles && this.props.output.allFiles.edges.length < 3 && this.props.output.allFiles.pageInfo.hasNextPage){
      this._loadMore();
    }
  }

  /*
    handle state and add listeners when component mounts
  */
  componentDidMount() {
    if(this.props.output.allFiles &&
      this.props.output.allFiles.pageInfo.hasNextPage) {
        this._loadMore()
    } else {
      this.setState({'moreLoading': false});
    }
  }

  /*
    @param
    triggers relay pagination function loadMore
    increments by 100
    logs callback
  */
  _loadMore() {
    this.setState({'moreLoading': true});
    let self = this;
    this.props.relay.loadMore(
     100, // Fetch the next 100 feed items
     (response, error) => {
       if(error){
         console.error(error)
       }

       if(self.props.output.allFiles &&
        self.props.output.allFiles.pageInfo.hasNextPage) {
          self._loadMore()
      } else {
        this.setState({'moreLoading': false})
      }
     }
   );
  }

  setRootFolder(key){
    this.setState({rootFolder: key})
    this._loadMore()
  }

  render(){

    this.props.loadStatus(this.state.moreLoading);
    if(this.props.output && this.props.output.allFiles){
      let outputFiles = this.props.output.allFiles
      if(this.props.output.allFiles.edges.length === 0){
        outputFiles = {
          edges: [],
          pageInfo: this.props.output.allFiles.pageInfo
        }
      }
      return(
        <FileBrowserWrapper
          ref="OutputBrowser"
          files={outputFiles}
          section="output"
          selectedFiles={this.props.selectedFiles}
          clearSelectedFiles={this.props.clearSelectedFiles}
          parentId={this.props.outputId}
          favoriteConnection="OutputFavorites_favorites"
          connection="OutputDataBrowser_allFiles"
          owner={this.props.owner}
          isLocked={this.props.isLocked}
          {...this.props}
        />
      )
    }else {
      return(<div>No Files Found</div>)
    }
  }
}


export default createPaginationContainer(
  OutputDataBrowser,
  {

    output: graphql`
      fragment OutputDataBrowser_output on LabbookSection{
        allFiles(after: $cursor, first: $first)@connection(key: "OutputDataBrowser_allFiles", filters:[]){
          edges{
            node{
              id
              isDir
              isFavorite
              modifiedAt
              key
              size
            }
            cursor
          }
          pageInfo{
            hasNextPage
            hasPreviousPage
            startCursor
            endCursor
          }
        }

      }`
  },
  {
    direction: 'forward',
    getConnectionFromProps(props) {
      return props.output && props.output.allFiles
    },
    getFragmentVariables(prevVars, totalCount) {
      return {
        ...prevVars,
        count: totalCount,
      };
    },
    getVariables(props, {count, cursor}, fragmentVariables) {

      const {owner, labbookName} = store.getState().routes

      return {
        first: count,
        cursor,
        owner: owner,
        name: labbookName
      };
    },
    query: graphql`
      query OutputDataBrowserPaginationQuery(
        $first: Int
        $cursor: String
        $owner: String!
        $name: String!
      ) {
        labbook(name: $name, owner: $owner){
           id
           description
           output{
            # You could reference the fragment defined previously.
            ...OutputDataBrowser_output
          }
        }
      }
    `
  }
)
