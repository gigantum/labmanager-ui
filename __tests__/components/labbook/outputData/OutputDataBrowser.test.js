
import React, {Component} from 'react'
import renderer from 'react-test-renderer';
import {mount} from 'enzyme'
import OutputDataBrowser from 'Components/labbook/outputData/OutputDataBrowser';

import { DragDropContext } from 'react-dnd'
import HTML5Backend from 'react-dnd-html5-backend'

import json from './__relaydata__/OutputDataBrowser.json'

import relayTestingUtils from 'relay-testing-utils'
      const loadStatus = ()=>{

      }


      const clearSelectedFiles = ()=>{

      }

      const backend = (manager: Object) => {
          const backend = HTML5Backend(manager),
              orgTopDropCapture = backend.handleTopDropCapture;

          backend.handleTopDropCapture = (e) => {

              if(backend.currentNativeSource){
                orgTopDropCapture.call(backend, e);

               //backend.currentNativeSource.item.dirContent = getFilesFromDragEvent(e, {recursive: true}); //returns a promise
              }
          };

          return backend;
      }

      const fixtures = {
        labbook: json.data.labbook,
        labbookId: json.data.labbook.id,
        isLocked: false,
        selectedFiles: [],
        clearSelectedFiles,
        outputId: json.data.labbook.output.id,
        output: json.data.labbook.output,
        loadStatus
      }

      class OutputDataCompInstance extends Component{
        render(){
          return (relayTestingUtils.relayWrap(<OutputDataBrowser {...fixtures}/>, {}, json.data.labbook))
        }
      }
      const OutputDataBrowserComponent = DragDropContext(backend)(OutputDataCompInstance)

      describe('Test CodeBrowser', () => {

        it('snapshot renders', () => {
          const wrapper = renderer.create(
             <OutputDataBrowserComponent />
          );

          const tree = wrapper.toJSON()

          expect(tree).toMatchSnapshot()
        })


      })
