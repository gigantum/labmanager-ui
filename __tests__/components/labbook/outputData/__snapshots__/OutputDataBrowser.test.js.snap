// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CodeBrowser snapshot renders 1`] = `
<div
  className="Code flex flex-row justify-center"
  id="code"
>
  <div
    className="rendered-react-keyed-file-browser"
  >
    <div
      className="rendered-file-browser"
    >
      <div
        className="action-bar"
      >
        <input
          onChange={[Function]}
          placeholder="Filter files"
          type="search"
          value=""
        />
        <ul
          className="item-actions"
        >
          <li>
            <a
              className="btn btn-primary btn-sm"
              href="#"
              onClick={[Function]}
              role="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-folder-o"
              />
              Â Add Folder
            </a>
          </li>
        </ul>
      </div>
      <div
        className="files"
        id="OutputDataBrowser_allFiles"
      >
        <table
          cellPadding="0"
          cellSpacing="0"
        >
          <thead>
            <tr
              className="folder"
            >
              <th>
                File
              </th>
              <th />
              <th
                className="size"
              >
                Size
              </th>
              <th
                className="modified"
              >
                Last Modified
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              className="file FileBrowser__large-dropzone modified grey"
            >
              <td
                colSpan="5"
              >
                Drag and Drop files here
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    className="DetailPanel hidden"
  >
    <div
      className="DetailPanel--close"
      onClick={[Function]}
    >
      X
    </div>
    <p />
    <p />
  </div>
</div>
`;
